File: 68.Text-Justification.go
// https://leetcode.com/problems/text-justification/
//
// algorithms
// Hard (22.02%)
// Total Accepted:    86,263
// Total Submissions: 391,701
// beats 100.0% of golang submission

package leetcode

import (
	"bytes"
	"strings"
)

func fullJustify(words []string, maxWidth int) []string {
	var cntWidth int
	var wordTmp []string
	var res []string

	for _, word := range words {
		wordLen := len(word)
		if len(wordTmp)+wordLen+cntWidth > maxWidth {
			space := maxWidth - cntWidth
			slot := len(wordTmp) - 1

			if slot > 0 {
				division := space / slot
				spaceArr := make([]int, slot)
				for i := 0; i < slot; i++ {
					spaceArr[i] = division
				}
				remainder := space % slot
				if remainder > 0 {
					for i := 0; i < slot && remainder != 0; i++ {
						spaceArr[i]++
						remainder--
					}
				}
				sb := bytes.NewBufferString("")
				for i := 0; i < slot; i++ {
					sb.WriteString(wordTmp[i])
					for j := 0; j < spaceArr[i]; j++ {
						sb.WriteString(" ")
					}
				}
				sb.WriteString(wordTmp[slot])
				res = append(res, sb.String())
			} else {
				sb := bytes.NewBufferString("")
				sb.WriteString(wordTmp[0])
				for j := 0; j < space; j++ {
					sb.WriteString(" ")
				}
				res = append(res, sb.String())
			}

			wordTmp = []string{word}
			cntWidth = wordLen
		} else {
			wordTmp = append(wordTmp, word)
			cntWidth += wordLen
		}
	}

	sb := strings.Join(wordTmp, " ")
	for i := len(sb); i < maxWidth; i++ {
		sb += " "
	}
	res = append(res, sb)

	return res
}

File: 68.Text-Justification.java
// https://leetcode.com/problems/text-justification/
//
// algorithms
// Hard (23.16%)
// Total Accepted:    95,628
// Total Submissions: 412,929


class Solution {
    public List<String> fullJustify(String[] words, int maxWidth) {
        ArrayList<String> l = new ArrayList<>();
        LinkedList<String> tmp = new LinkedList<>();
        
        int wordSum = 0, numSum = 0;
        
        for (String w : words) {
            int wLen = w.length();
            if (wordSum + numSum + wLen <= maxWidth) {
                wordSum += wLen;
                numSum++;
                tmp.offer(w);
            } else {
                StringBuilder sb = new StringBuilder();
                if (numSum == 1) {
                    sb.append(tmp.peek());
                    for (int i = 0; i < maxWidth - wordSum; i++) {
                        sb.append(" ");
                    }
                } else {
                    int[] space = new int[numSum - 1];
                    int spaceSum = maxWidth - wordSum;
                    Arrays.fill(space, spaceSum / (numSum - 1));
                    for (int i = 0; i < spaceSum % (numSum - 1); i++) {
                        space[i]++;
                    }
                    for (int i = 0; i < numSum - 1; i++) {
                        sb.append(tmp.pollFirst());
                        for (int j = 0; j < space[i]; j++) {
                            sb.append(" ");
                        }
                    }
                    sb.append(tmp.peek());
                }
                l.add(sb.toString());
                wordSum = wLen;
                numSum = 1;
                tmp.clear();
                tmp.offer(w);
            }
        }
        
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < numSum - 1; i++) {
            sb.append(tmp.pollFirst());
            sb.append(" ");
        }
        sb.append(tmp.peek());
        for (int i = 0; i < maxWidth - wordSum - (numSum - 1); i++) {
            sb.append(" ");
        }
        l.add(sb.toString());        
        
        return l;
    }
}